### DISCUSSION TITLE
Identity-Aware Proxy

### DISCUSSION THREAD

**Discussion by electroheadfx** [nextauthjs/next-auth#5365] [Ideas]
### Description 📓

is exist an implementation with Google Identity-Aware Proxy (IAP) ? I'm not sure to find it in the doc. I think its totally different implementation with Google Auth2 identity

### How to reproduce ☕️

https://cloud.google.com/iap/docs/enable-external-identities

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by kriswuollett** [comment]
@electroheadfx there doesn't appear to be any implementation for Google IAP which I'd imagine being called a `jwt` auth type. NextAuth only has `oauth`, `email`, and `credentials`. If you take a look at the `credentials`' `authorize` callback you'll see that there is `headers` available:

https://github.com/nextauthjs/next-auth/blob/0e6c51adecc54d3102ef3cda57f909400bc99c29/packages/next-auth/src/core/routes/callback.ts#L325-L329

I'm interested in this as well, and I'd probably approach it as creating a custom `credentials` provider with a `jwt` field, use a custom `signIn` page parameter that is handled in middleware that posts to the callback which in turn should let the provider verify the JWT:

https://github.com/nextauthjs/next-auth/blob/45f423ed5d9349ee1b8c5574ff6153053d894361/packages/core/src/lib/routes/callback.ts#L261-L265

And if verified the middleware should then redirect the user to the originally requested URL after optionally using the [People API](https://developers.google.com/people) to add details like photos beforehand.

@balazsorban44, perhaps something like a `jwt` provider type could be created to avoid writing a workaround-like custom middleware? Could take in the parameters like the `jose.JWTVerifyGetKey` and header name to find the JWT.

**Reply by garnerp** [reply]
@kriswuollett did you ever end up making this work?

**Reply by kriswuollett** [reply]
@garnerp no, I haven't yet, other project priorities put this in the back of the queue.

**Comment by aryzle** [comment]
I opted to follow the example here https://github.com/vercel/examples/tree/main/edge-middleware/jwt-authentication instead of using NextAuth

**Reply by garnerp** [reply]
I see a lot of value in taking advantage of using next-auth's session/jwt capabilities and the client hooks and server side getServerSession.  Did you end up building all of that in yourself?  I can see the above example would be sufficient to stop anyone from getting into your application's protected routes based on a valid JWT.

**Reply by aryzle** [reply]
I did not see much value in next-auth's session/jwt capabilities for our use case, I just pass the the user info we need to the client through a cookie, otherwise things happen server side. Also passing an exp time to the client so we can check it and if expired force a login.

What pieces of the session/jwt capabilities do you see as most valuable? hopefully I'm not missing something 😄
